### PR TITLE
RefMan: Adding draft material about modules

### PR DESCRIPTION
**Pull request by davidcok**

### CODE REVIEW

File: docs/_includes/Grammar.md [hunk 1]
```diff
@@ -98,9 +98,8 @@ Dafny source code files are readable text encoded as UTF-8 Unicode
 All program text other than the contents of comments, character, string and verbatim string literals
 are printable and white-space ASCII characters,
 that is, ASCII characters in the range `!` to `~`, plus space, tab, cr and nl (ASCII, 9, 10, 13, 32)  characters,
```
Comment (RustanLeino) [inline docs/_includes/Grammar.md:100]:
```suggestion
that is, ASCII characters in the range `!` to `~`, plus space, tab, cr and nl (ASCII, 9, 10, 13, 32)  characters.
```

File: docs/_includes/Modules.md [hunk 2]
```diff
@@ -15,12 +15,12 @@ interface.
 ## 4.1. Declaring New Modules
 ````grammar
 ModuleDefinition_ = "module" { Attribute } ModuleName
-        [ [  "exclusively" ] "refines" ModuleQualifiedId ]
+        [ [ "exclusively" ] "refines" QualifiedModuleName ]
```
Comment (RustanLeino) [inline docs/_includes/Modules.md:18]:
```suggestion
        [ "refines" QualifiedModuleName ]
```

The `exclusively` was dropped several years ago.

File: docs/_includes/Modules.md [hunk 3]
```diff
@@ -648,6 +649,75 @@ Ordinarily a module must be _imported_ in order for its constitutent
...
+the imported names in the refinement parent.
+
+Within each namespace, the local names are unique. Thus a module may
+not reuse a name that a refinement parent has declared (unless it is a
+refining declaration, which replaces both declarations, as described
+[here](#TODO)).
+
+Local names take precedence over imported names. If a name is used more than
+once among imported names (coming from different imports), then it is
+ambiguous to _use_ the name without qualification.
```
Comment (RustanLeino) [inline docs/_includes/Modules.md:682]: unless they refer to types and those types are equal.

File: docs/_includes/Modules.md [hunk 4]
```diff
@@ -648,6 +649,75 @@ Ordinarily a module must be _imported_ in order for its constitutent
...
+0. The leading ``NameSegment`` is resolved as a local or imported module name of `Z`, if there
+is one with a matching name. The target of a `refines` clause does not 
+consider local names, that is, in `module Z refines A.B.C`, any contents of `Z`
+are not considered in finding `A`.
+
+1. Otherwise, it is resolved as a local or imported module name of the most enclosing module of `Z`,
+   iterating outward to each successive enclosing module until a match is
+found or the default toplevel module is reached without a match.
+No consideration of export sets, default or otherwise, is used in this step.
+Howecver, if at any stage a matching name is found that is not a module
```
Comment (RustanLeino) [inline docs/_includes/Modules.md:699]:
```suggestion
However, if at any stage a matching name is found that is not a module
```

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: docs/_includes/Grammar.md (hunk 1 above)
Comment (davidcok) [reply docs/_includes/Grammar.md:100]: OK

File: docs/_includes/Modules.md (hunk 2 above)
Comment (davidcok) [reply docs/_includes/Modules.md:18]: OK

File: docs/_includes/Modules.md (hunk 3 above)
Comment (davidcok) [reply docs/_includes/Modules.md:682]: Right

File: docs/_includes/Modules.md (hunk 4 above)
Comment (davidcok) [reply docs/_includes/Modules.md:699]: OK

**Pull request merged**